binlog_reconciler.go

```diff
@@ -0,0 +1,223 @@
+// INSERT or UPDATE occured. For a DELETE statement, it generates the
+// corresponding DELETE statement based on the primary key.
+//
+// This is used during the resume when we are catching up to the binlog events
+// missed by Ghostferry while it is down.
+type ReconcilationDMLEvent struct {
+	*DMLEventBase
+	newValues RowData
+	pk        uint64
+}
+
+func (e *ReconcilationDMLEvent) OldValues() RowData {
+	return nil
+}
+
+func (e *ReconcilationDMLEvent) NewValues() RowData {
+	return e.newValues
+}
+
+func (e *ReconcilationDMLEvent) PK() (uint64, error) {
+	return e.pk, nil
+}
+
+func (e *ReconcilationDMLEvent) AsSQLString(target *schema.Table) (string, error) {
+	var query string
+	if e.newValues != nil {
+		columns, err := loadColumnsForTable(&e.table, e.newValues)
+		if err != nil {
+			return "", err
+		}
+
+		query = "REPLACE INTO " +
+			QuotedTableNameFromString(target.Schema, target.Name) +
+			" (" + strings.Join(columns, ",") + ")" +
+			" VALUES (" + buildStringListForValues(e.newValues) + ")"
+	} else {
+		pkColumnName := e.TableSchema().GetPKColumn(0).Name
+		if pkColumnName == "" {
+			return "", fmt.Errorf("cannot get PK column for table %s", e.Table())
+		}
+
+		pkColumnName = quoteField(pkColumnName)
+		query = "DELETE FROM " + QuotedTableNameFromString(target.Schema, target.Name) +
+			" WHERE " + buildStringMapForWhere([]string{pkColumnName}, []interface{}{e.pk})
+	}
+
+	return query, nil
+}
+
+func NewReconciliationDMLEvent(table *schema.Table, pk uint64, row RowData) DMLEvent {
+	return &ReconcilationDMLEvent{
+		DMLEventBase: &DMLEventBase{table: *table},
+		pk:           pk,
+		newValues:    row,
+	}
+}
+
+// Instead of replacing/deleting every row, we first store all the rows changed
+// during the downtime and perform all the operations at the end. This maybe
+// faster (trading memory usage tho) and it is crucial for schema changes (as
+// we can simply delete a key from this map when we realize a table has
+// changed).
+type UniqueRowMap map[TableIdentifier]map[uint64]struct{}
+
+func (m UniqueRowMap) AddRow(table *schema.Table, pk uint64) bool {
+	tableId := NewTableIdentifierFromSchemaTable(table)
+	if _, exists := m[tableId]; !exists {
+		m[tableId] = make(map[uint64]struct{})
+	}
+
+	if _, exists := m[tableId][pk]; !exists {
+		m[tableId][pk] = struct{}{}
+		return true
+	}
+
+	return false
+}
+
+type BinlogReconciler struct {
+	BatchSize        int
+	TableSchemaCache TableSchemaCache
+
+	SourceDB     *sql.DB
+	BinlogWriter *BinlogWriter
+
+	modifiedRows UniqueRowMap
+	logger       *logrus.Entry
+}
+
+func (r *BinlogReconciler) Initialize() {
+	r.modifiedRows = make(UniqueRowMap)
+	r.logger = logrus.WithField("tag", "binlogreconcile")
+}
+
+func (r *BinlogReconciler) AddRowsToStore(events []DMLEvent) error {
+	for _, ev := range events {
+		pk, err := ev.PK()
+		if err != nil {
+			return err
+		}
+
+		r.modifiedRows.AddRow(ev.TableSchema(), pk)
```

Comment:
do we really wanna `REPLACE` for every table? seems like the usual binlog playing logic should handle most tables?